### PR TITLE
PE-1086: Progress callback on tx uploader

### PR DESCRIPTION
**Pull request by fedellen**
This PR adds a debounced progress callback to the concurrent tx uploader which can currently be enabled with an ENV variable, similar to the persistent cache pattern

### CODE REVIEW

**Comment by fedellen** [comment]
Notably, this minimalistic approach doesn't account for multiple transactions coming in from `uploadAllEntities` and only reports progress on one transaction at a time.

Will need to modify the overlying method to include that info

File: src/arfs/arfsdao.ts [hunk 1]
```diff
@@ -1063,16 +1095,40 @@ export class ArFSDAO extends ArFSDAOAnonymous {
...
+					// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+					this.shouldProgressLog && transaction.chunks!.chunks.length > defaultMaxConcurrentChunks;
+
 				const transactionUploader = new MultiChunkTxUploader({
 					transaction,
-					gatewayUrl: gatewayUrlForArweave(this.arweave)
+					gatewayUrl: gatewayUrlForArweave(this.arweave),
+					progressCallback: shouldProgressLog
+						? (pct: number) => {
+								if (!debounce) {
```
Comment (arielmelendez) [inline src/arfs/arfsdao.ts:1115]: Minor UX improvement might be to allow the 100% message to always come through.

File: src/arfs/arfsdao.ts [hunk 2]
```diff
@@ -1063,16 +1095,40 @@ export class ArFSDAO extends ArFSDAOAnonymous {
 	}
 
 	async sendTransactionsAsChunks(transactions: Transaction[]): Promise<void> {
+		let debounce = false;
```
Comment (arielmelendez) [inline src/arfs/arfsdao.ts:1098]: nit: only relevant if !dryRun. Might be better to improve locality of reference by moving in that if block.

File: src/arfs/arfsdao.ts [hunk 3]
```diff
@@ -1063,16 +1095,40 @@ export class ArFSDAO extends ArFSDAOAnonymous {
...
+									}, 250); // .25 sec debounce
+								}
+						  }
+						: undefined
 				});
 
 				await transactionUploader.batchUploadChunks();
+
+				if (shouldProgressLog) {
+					console.info('Transaction upload complete!');
```
Comment (arielmelendez) [inline src/arfs/arfsdao.ts:1130]: Could be nice to have the txId in here.

File: src/arfs/arfsdao.ts [hunk 4]
```diff
@@ -1063,16 +1095,40 @@ export class ArFSDAO extends ArFSDAOAnonymous {
...
+					this.shouldProgressLog && transaction.chunks!.chunks.length > defaultMaxConcurrentChunks;
+
 				const transactionUploader = new MultiChunkTxUploader({
 					transaction,
-					gatewayUrl: gatewayUrlForArweave(this.arweave)
+					gatewayUrl: gatewayUrlForArweave(this.arweave),
+					progressCallback: shouldProgressLog
+						? (pct: number) => {
+								if (!debounce) {
+									console.info(`Transaction Upload Progress: ${pct}%`);
```
Comment (arielmelendez) [inline src/arfs/arfsdao.ts:1116]: Could be nice to have the txId in here.

File: src/arfs/multi_chunk_tx_uploader.ts [hunk 5]
```diff
@@ -96,6 +100,10 @@ export class MultiChunkTxUploader {
 		this.transaction = transaction;
 		this.maxConcurrentChunks = maxConcurrentChunks;
 		this.maxRetriesPerRequest = maxRetriesPerRequest;
+
+		if (progressCallback) {
```
Comment (arielmelendez) [inline src/arfs/multi_chunk_tx_uploader.ts:104]: Is this check necessary?

File: src/arfs/multi_chunk_tx_uploader.ts [hunk 6]
```diff
@@ -151,6 +160,10 @@ export class MultiChunkTxUploader {
 			}
 
 			this.uploadedChunks++;
+
+			if (this.progressCallback) {
```
Comment (arielmelendez) [inline src/arfs/multi_chunk_tx_uploader.ts:164]: You can use optional chaining on functions:
```
this.progressCallback?.(this.pctComplete);
```

**Review by arielmelendez — CHANGES_REQUESTED** [review]
A few minor comments.

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**